### ISSUE TITLE
GRPC streaming hanging after first or second cancelled request

### ISSUE THREAD

**Issue by esbie** [linkerd/linkerd#1008]
To reproduce:
checkout `esbie/grpc` branch from https://github.com/BuoyantIO/linkerd-examples
```
cd gob
docker-compose build
docker-compose up -d
curl -v localhost:4141/gob
```

Then ctrl-c. It takes one or two of these cancelled requests, and then all subsequent streaming requests will hang. I'm able to verify streaming works correctly if the services talk directly to each other.

reproducing the linkerd config here:
```
routers:
- protocol: http
  servers:
  - port: 4141
    ip: 0.0.0.0
  baseDtab: |
    /srv => /$/inet/web/8080;
    /http/*/* => /srv;

- protocol: h2
  experimental: true
  label: grpc
  servers:
  - port: 4142
    ip: 0.0.0.0
  dstPrefix: /grpc
  identifier:
    kind: io.l5d.headerPath
  baseDtab: |
    /srv/proto.GenSvc   => /$/inet/gen/8181;
    /srv/proto.WordSvc  => /$/inet/word/8282;
    /grpc => /srv;
```

unhelpful linkerd logs:
```
E 0127 00:54:42.991 THREAD25 TraceId:b7797fc2decb4aea: service failure
com.twitter.finagle.CancelledRequestException: request cancelled. Remote Info: Upstream Address: /10.240.0.19:55029, Upstream Client Id: Not Available, Downstream Address: /127.0.0.1:8282, Downstream Client Id: $/inet/127.1/8282, Trace Id: b7797fc2decb4aea.02a8763bed0d8ecb<:b7797fc2decb4aea
```

unhelpful gob logs:
```
2017/01/27 18:18:08 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:5252: getsockopt: connection refused"; Reconnecting to {localhost:5252 <nil>}
2017/01/27 18:18:08 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:5252: getsockopt: connection refused"; Reconnecting to {localhost:5252 <nil>}
2017/01/27 18:18:11 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:5252: getsockopt: connection refused"; Reconnecting to {localhost:5252 <nil>}
2017/01/27 18:18:11 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:5252: getsockopt: connection refused"; Reconnecting to {localhost:5252 <nil>}
2017/01/27 18:18:15 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:5252: getsockopt: connection refused"; Reconnecting to {localhost:5252 <nil>}
2017/01/27 18:18:15 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:5252: getsockopt: connection refused"; Reconnecting to {localhost:5252 <nil>}
write error: write tcp 127.0.0.1:8282->127.0.0.1:56650: write: broken pipe
```

The linkerd used is a snapshot from master after https://github.com/linkerd/linkerd/pull/1003 was merged.

**Comment by ondrej-smola** [comment]
Same here, using go grpc client/server. Cancelling client hangs all following requests. 
From what i observed after first cancel: Client -> Linkerd -> Server. 
C opens connection and sends M1 to S through L, M1 arrives at S and S reply with M2. M2 is send but never arrives at C. C hangs at recv and S hangs at recv. 

First request

```
0205 19:44:57.645 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] ServiceName(#/io.l5d.marathon/test-services/v2t-engine-proxy)
0205 19:44:57.645 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] BinaryAnnotation(clnt/finagle.version,6.41.0)
0205 19:44:57.645 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] ClientSend()
0205 19:44:57.646 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] WireSend
0205 19:44:57.646 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] ServerAddr(/192.168.91.61:21716)
D 0205 19:44:57.646 THREAD57 TraceId:9ec900a6de5a7faf: [C L:/172.17.0.3:44066 R:/192.168.91.61:21716 S:3] initialized stream
0205 19:44:57.647 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] WireRecv
0205 19:44:57.648 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] BinaryAnnotation(l5d.success,1.0)
0205 19:44:57.648 9ec900a6de5a7faf.9ec900a6de5a7faf<:9ec900a6de5a7faf] ClientRecv()
0205 19:44:57.648 b926bc1d3375b04d.b926bc1d3375b04d<:b926bc1d3375b04d] ServerSend()
D 0205 19:45:02.082 THREAD76: [S L:/172.17.0.3:8080 R:/192.168.91.56:39734 S:11] stream reset from remote: Reset.NoError
D 0205 19:45:02.083 THREAD76 TraceId:9ec900a6de5a7faf: [C L:/172.17.0.3:44066 R:/192.168.91.61:21716 S:3] stream read failed: Reset.NoError
Reset.NoError

D 0205 19:45:02.083 THREAD76 TraceId:9ec900a6de5a7faf: [C L:/172.17.0.3:44066 R:/192.168.91.61:21716 S:3] resetting Reset.NoError in Open(com.twitter.finagle.buoyant.h2.netty4.Netty4StreamTransport$RemoteStreaming@19b5eba4)
D 0205 19:45:02.083 THREAD76 TraceId:9ec900a6de5a7faf: [C L:/172.17.0.3:44066 R:/192.168.91.61:21716 S:3] stream closed
D 0205 19:45:02.083 THREAD76: [S L:/172.17.0.3:8080 R:/192.168.91.56:39734 S:11] remote write failed: Reset.NoError
Reset.NoError

```

Second
```
0205 19:45:27.821 b926bc1d3375b04d.b926bc1d3375b04d<:b926bc1d3375b04d] Message(namer.success)
0205 19:45:27.821 b926bc1d3375b04d.b926bc1d3375b04d<:b926bc1d3375b04d] BinaryAnnotation(dst.id,/#/io.l5d.marathon/test-services/v2t-engine-proxy)
0205 19:45:27.821 b926bc1d3375b04d.b926bc1d3375b04d<:b926bc1d3375b04d] BinaryAnnotation(dst.path,/)
0205 19:45:27.822 9450aad50b1abbee.9450aad50b1abbee<:9450aad50b1abbee] ServiceName(#/io.l5d.marathon/test-services/v2t-engine-proxy)
0205 19:45:27.822 9450aad50b1abbee.9450aad50b1abbee<:9450aad50b1abbee] BinaryAnnotation(clnt/finagle.version,6.41.0)
0205 19:45:27.822 9450aad50b1abbee.9450aad50b1abbee<:9450aad50b1abbee] ClientSend()
0205 19:45:27.822 9450aad50b1abbee.9450aad50b1abbee<:9450aad50b1abbee] WireSend
0205 19:45:27.822 9450aad50b1abbee.9450aad50b1abbee<:9450aad50b1abbee] ServerAddr(/192.168.91.61:21716)
D 0205 19:45:27.822 THREAD62 TraceId:9450aad50b1abbee: [C L:/172.17.0.3:44066 R:/192.168.91.61:21716 S:9] initialized stream
```

Then it hangs.

If somebody can lead me how to debug this or point me to relevant parts of project.
We are  building prototype and this is show stopper for us.

**Comment by wmorgan** [comment]
There's a rumor we fixed some of this in 0.9.0. Is this still reproducible?

**Comment by esbie** [comment]
not reproducible on buoyantio/linkerd:0.9.1
